skglm/solvers/prox_newton.py

```diff
@@ -0,0 +1,338 @@
+
+        for epoch in range(max_epochs):
+
+            # find descent direction
+            if is_sparse:
+                delta_w_ws, X_delta_w_ws = _descent_direction_s(
+                    *X_bundles, y, w, Xw, grad_ws, datafit,
+                    penalty, ws, tol=EPS_TOL*tol_in)
+            else:
+                delta_w_ws, X_delta_w_ws = _descent_direction(
+                    X, y, w, Xw, grad_ws, datafit, penalty, ws, tol=EPS_TOL*tol_in)
+
+            # backtracking line search with inplace update of w, Xw
+            if is_sparse:
+                grad_ws[:] = _backtrack_line_search_s(*X_bundles, y, w, Xw,
+                                                      datafit, penalty, delta_w_ws,
+                                                      X_delta_w_ws, ws)
+            else:
+                grad_ws[:] = _backtrack_line_search(X, y, w, Xw, datafit, penalty,
+                                                    delta_w_ws, X_delta_w_ws, ws)
+
+            # check convergence
+            opt_in = penalty.subdiff_distance(w, grad_ws, ws)
+            stop_crit_in = np.max(opt_in)
+
+            if max(verbose-1, 0):
+                p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+                print(
+                    f"Epoch {epoch+1}: {p_obj:.10f}, "
+                    f"stopping crit in: {stop_crit_in:.2e}"
+                )
+
+            if stop_crit_in <= tol_in:
+                if max(verbose-1, 0):
+                    print("Early exit")
+                break
+
+        p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+        p_objs_out.append(p_obj)
+    return w, p_objs_out, stop_crit
+
+
+@njit
+def _descent_direction(X, y, w_epoch, Xw_epoch, grad_ws, datafit,
+                       penalty, ws, tol):
+    # Given:
+    #   1) b = \nabla F(X w_epoch)
+    #   2) D = \nabla^2 F(X w_epoch)  <------>  raw_hess
+    # Minimize quadratic approximation for delta_w = w - w_epoch:
+    #  b.T @ X @ delta_w + \
+    #  1/2 * delta_w.T @ (X.T @ D @ X) @ delta_w + penalty(w)
+    raw_hess = datafit.raw_hessian(y, Xw_epoch)
+
+    lipschitz = np.zeros(len(ws))
+    for idx, j in enumerate(ws):
+        lipschitz[idx] = raw_hess @ X[:, j] ** 2
+
+    cached_grads = np.zeros(len(ws))
+    X_delta_w_ws = np.zeros(X.shape[0])
+    w_ws = w_epoch[ws]
+
+    for cd_iter in range(MAX_CD_ITER):
+        for idx, j in enumerate(ws):
+            # skip when X[:, j] == 0
+            if lipschitz[idx] == 0:
+                continue
+
+            cached_grads[idx] = grad_ws[idx] + X[:, j] @ (raw_hess * X_delta_w_ws)
+            old_w_idx = w_ws[idx]
+            stepsize = 1 / lipschitz[idx]
+
+            w_ws[idx] = penalty.prox_1d(
+                old_w_idx - stepsize * cached_grads[idx], stepsize, j)
+
+            if w_ws[idx] != old_w_idx:
+                X_delta_w_ws += (w_ws[idx] - old_w_idx) * X[:, j]
+
+        if cd_iter % 5 == 0:
+            opt = penalty.subdiff_distance(w_ws, cached_grads, ws)
```

Comment:
so the optimality condition is approximated (to avoid recomputing the gradients fully); but the gradients are not exact (all of them but 1 are from older updates). Right?

Reply:
no, it's not
refer to https://github.com/scikit-learn-contrib/skglm/pull/51#discussion_r952329694

Reply:
every time we update Xw, the gradient wrt any feature changes, no?
So if you compute gradient wrt to feature 1, update feature, compute gradient wrt feature 2, update feat 2, then hasn't the gradient for feature 1?